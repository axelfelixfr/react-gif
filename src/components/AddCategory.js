import React, { useState } from 'react';
import PropTypes from 'prop-types';

export const AddCategory = ({setCategories}) => {

    // Le damos un valor por defecto al input con el useState
    const [inputValue, setInputValue] = useState(''); // Mandamos como estado por defecto un string vacio

    // Realizamos una función para el input
    const handleInputChange = (e) => {
        // Hacemos uso del método setInputValue del useState para cambiar su value
        setInputValue(e.target.value);
    };

    const handleSubmit = (e) => {
        e.preventDefault(); // Quitamos el evento por default
        // Verificamos que por lo menos haya un string con longitud de tres para no mandar datos vacíos
        if(inputValue.trim().length > 2){
            // Accedemos a las categorías y le agregamos lo que esta en inputValue
            setCategories(cats => [inputValue, ...cats]);
            // Borramos el contenido del input
            setInputValue('');
        }
    };

    return (
        // Le pasamos el evento handleSubmit al formulario
        <form onSubmit={handleSubmit}>
            <input 
                type="text"
                // Le pasamos el value del useState
                value={ inputValue }
                // Le pasamos la función que se ejecutara cada que cambie el input
                onChange={ handleInputChange }
            />
        </form>
    )
}

// setCategories pasa a ser requerida
AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired
}