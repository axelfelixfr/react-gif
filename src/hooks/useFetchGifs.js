import { useEffect, useState } from 'react';
import {getGifs} from '../helpers/getGifs';

// Realizamos un hook para la API
export const useFetchGifs = (category) => {
    const [state, setState] = useState({
        data: [],
        loading: true
    });
    
    // Los useEffect no pueden ser asincronos, ya que son sincronos
    useEffect(() => {
        // Obtenemos los gifs a través de la categoría y cambiamos su estado con setState
        getGifs(category).then(imgs => {
            // Le pasamos la data de los gifs al igual que pasamos el loading de true a false, ya que termino la carga
            setState({
                data: imgs,
                loading: false
            });
        });
    }, [category]); // Si cambia la categoría, cambiara el useEffect

    return state; // { data:[], loading: true }
}