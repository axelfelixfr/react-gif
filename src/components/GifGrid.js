import React from 'react';
import PropTypes from 'prop-types';
import { useFetchGifs } from '../hooks/useFetchGifs';
import { GifGridItem } from './GifGridItem';

export const GifGrid = ({ category }) => {
    // A la data la renombramos images
    const { data:images, loading } = useFetchGifs(category);

    return (
        <>
            {/* Le pasamos el nombre de la categoría (de la búsqueda que se hizo) */}
            <h3 className="animate__animated animate__fadeIn">{category}</h3>
            
            {/* El loading debe estar en true para que aparezca el parrafo */}
            { loading && <p className="animate__animated animate__flash">Loading</p> }
            <div className="card-grid">
            
            {/* Extraemos todos los gifs con map, y le pasamos al componente <GifGridItem /> la data de cada gif que se recoge */}
                {
                    images.map(img => (
                        <GifGridItem 
                            key={ img.id }
                            { ...img }
                        />
                    ))              
                }
            
            </div>
        </>
    )
}

GifGrid.propTypes = {
    category: PropTypes.string.isRequired
}