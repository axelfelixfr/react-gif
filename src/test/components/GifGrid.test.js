import React from 'react';
import '@testing-library/jest-dom';
import { shallow } from 'enzyme';
import { GifGrid } from '../../components/GifGrid';
import { expect } from '@jest/globals';
import { useFetchGifs } from '../../hooks/useFetchGifs';
jest.mock('../../hooks/useFetchGifs');

describe('Pruebas en el componente <GifGrid />', () => {
    
    const category = 'FUN';

    test('Debe de cargar correctamente el componente', () => {
        useFetchGifs.mockReturnValue({
            data: [],
            loading: true
        });

        const wrapper = shallow(<GifGrid category={category}/>);

        expect(wrapper).toMatchSnapshot();
    })

    test('Debe de mostrar items cuando se cargan imÃ¡genes useFetchGifs', () => {
        const gifs = [{
            id: 'ABC',
            url: 'https://localhost/ejemplo/prueba.jpg',
            title: 'Ejemplo de prueba'
        },
        {
            id: 'Nuevo',
            url: 'https://localhost/ejemplo/nuevo.jpg',
            title: 'Ejemplo nuevo'
        }];

        useFetchGifs.mockReturnValue({
            data: gifs,
            loading: false
        });

        const wrapper = shallow(<GifGrid category={category}/>);

        // expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('p').exists()).toBe(false);
        expect(wrapper.find('GifGridItem').length).toBe(gifs.length);
    })
    
    
})