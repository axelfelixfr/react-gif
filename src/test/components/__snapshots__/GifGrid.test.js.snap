// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente <GifGrid /> Debe de cargar correctamente el componente 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__fadeIn"
  >
    FUN
  </h3>
  <p
    className="animate__animated animate__flash"
  >
    Loading
  </p>
  <div
    className="card-grid"
  />
</Fragment>
`;
